packages/engine/Source/Scene/GlobeSurfaceShaderSet.js

```diff
@@ -10,13 +10,15 @@ function GlobeSurfaceShader(
   flags,
   material,
   shaderProgram,
-  clippingShaderState
+  clippingShaderState,
+  vertexShadowDarkness
```

Comment:
The `GlobeSurfaceShaderSet` is used when new shader variants need to be generated. Since we're just using a `uniform` for the `vertexShadowDarkness`, I believe this file should not be changed at all.

Reply:
@sanjeetsuhag thanks for looking into this. The purpose for making the GlobeSurfaceShaderSet aware of the `vertexShadowDarkness` is for dynamically changing that property value at run time, as it is visible in the last sandcastle example I shared.
If the scene gets pretty dark (that's the cost of having a more realist terrain), the user might want to decrease the darkness for better exploring the terrain.

In the code, the `vertexShadowDarkness` is passed to the `GlobeSurfaceShaderSet`. There it is checked if it has changed and if it has then a redraw is forced. What do you think? Are there other ways to refresh the terrain when the `vertexShadowDarkness` changes?

Reply:
> In the code, the vertexShadowDarkness is passed to the GlobeSurfaceShaderSet. There it is checked if it has changed and if it has then a redraw is forced. What do you think? Are there other ways to refresh the terrain when the vertexShadowDarkness changes?

Any changes to the  `Globe.vertexShadowDarkness` are propagated to the `GlobeSurfaceTileProvider`, where the value of the `u_vertexShadowDarkness` uniform is updated for use in the shader.

https://github.com/calogeromauceri/cesium/blob/bcd347c8bd31a5049f7040d672421f007599b7c8/packages/engine/Source/Scene/GlobeSurfaceTileProvider.js#L1789-L1791

This is all that is needed to provide an updated value for a `uniform` to shader. For cases where we need to alter the shader code itself, that's when a new shader variant should be compiled.  We mainly use preprecessor directives for these. For example, then the globe vertex lighting is enabled in the scene, the `#define ENABLE_VERTEX_LIGHTING` mandates a recompilation of the shader. For dynamic quantities like `vertexShadowDarkness`, it is preferable to provide them as `uniform`.

I removed all the changes from `GlobeSurfaceShaderSet` in your branch, and created this [local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=lVZtb9s2EP4rhL9EAWxaSpy1c5xgazO0GZq1WNIW2DwUtHS2uFCkRlJ23SL/fXeiJNuaO2xOAJH38vDeybleC8tAKVk6IzN2xTRs2Etwsir4Ty05Sp6dP5vE8bD3Pb2c67nuS/OPr+6fTxCqg0UxOqYQ5Ttr/oTUS6MPj3oFZmVFmct0JxF1+uEcglgps4CeKpEOZWsp7nKzeWVNpbMffWFcmYMl1aVQDi7ZeIwgKTgn7JYtjWVeKnDMG0Z6TGqmjR6BsD7fOdJCgxYLBW/kKvdSrxDU2wpqIxH2IQe2lrC5Mxko5Ij0EXGR6LzwwMySmQqDIbVkoiyVTAV5y+c6Ndr5PdUr9nWuGVuD9fD5PheZ2dwI+6jR6GmIBD/GG87102VtyUujSaA+fAdbQLEA69BFdPZRm/TRVJ6ZhQO7Jrcc73LacnntRdRhnLa+vpA668EjKBFu3t5h3KAA7R35TLT3t/gRnqVCqTrm0ndue2PUAjN8xTKTVqTGV+B/CggvtrdZNB80MvMBnd83kWK5JXswI25n6rBFDjYvKx3Kz1ULl1q5gHfCigI82Ejj9zTEvAdOJMbIoLddmPaPqDUboQ446s6KsFw/CFW16PSr8/c7Kf4R6rkWuAzsJzL2iew9YuZ8cCzrISjhjzoFJa2QGttpLTOwhz0TPg+HIlFjW2XVlJ3k3pduOh4vqS24SP0Ki5enphg3yGNVaWE/rVSWxPHJMOh2nTLdLRuWhb8qcP5DbfsvxhbYh9O6cdDVXYebEuvlqo3TwciYHm4b3DqQTT80pJ7v0z6hEXOP291kmIbB0LLq2LodMZCx4l8jAKv0bna8v21UsNpru26xBLGnjT2EXAgHb8QW7DuZPkKP6WUBSmoIAWk9A5OarC8q9dK8MJ979qaggYrxGLjQshAhgC35qYs3VXG/PD7UtOgkrbc4RTwGD+zJsM4O5Spo8fpYvjSrZiJm3XwN4HjuCuy2dpuy2qgdkFH0YM9Flt0GQleZe7Z9hMWdKO+xCWUKfbldBbO9Et7AAguHK2tSLlzFIavGbcGqYBpKUzl/2oj0E1WSUK1A2fYdCXXdy2iua4dMnE7sIGv0W1J5E/2kdm1c6tVJ09ldwjHdq/s0x/k2PX4TPuyJRF+PthYOih0efJSZz6fsIjnbI74GuqVqKvVZOySwku8hXDZsI/GOy0HQ5ByyUvo0x5rJmDVKdalOMQpWcAeeqiNEOsN+lroprXa04I2JK6HP+dKa4gZWFsBFbWxGScIvki5Uo8k5/34y6fbJ+Tk+L9rd8ZdF4AbHjZWYgMaCJjuNJ51Fd8Ln3JtfkYopi5IJv2ijVvv6LcHR5Iw/byUpFt8SjHkcxCi5T22E27gpvEI4Xv3WP2CTYxM0ID9XCrVvaKpSoG6def5dnOB0P4vPklF8MTqbPCSTaRzjP4/j+Lcw4Xuwpvz/qGfJv6OmlaWyboCP+7HTaUcX/2JM8WCi4/JDdtTw08vBcDBzfqvgus36D7IoDb5asIsjzvG6KUqF/rjxosLZ5nnq6hE0G7dqs0yumcyu5oPewJoPWKqEc8hZVkrdyy8wH1zPxih/oKZMXTFv8V7FeUAieXL9JhA557Mxbv+p1b1GiOpp/NUuzPzCZNvGm5m3nV8zn12H24+Fq5u1d/dsjKx9sd0kmUld4uvMb0vAI3HmrNAFejziDhNIa/EZ10m9dh7KwEhwlwkvRgt8ESFpTa+L6dG3JKaGmO9LlAcaWXTiSe3WcSsQwdNhGE7cTf77SXuYey7jMgQJF03ocNXEsw07Cf4N) for demonstration:

![vertexShadowDarkness](https://user-images.githubusercontent.com/5172619/205685819-a194738e-3197-4eba-bf5a-bb0963666005.gif)

Reply:
@sanjeetsuhag thanks for the explanation. It is much clearer to me now. I made the cleanup you suggested and indeed it is working as expected